### DISCUSSION TITLE
ImageMagick on Azure Functions cannot load native library

### DISCUSSION THREAD

**Discussion by madushans** [dlemstra/Magick.NET#689] [Help] [answered]
**Note:** I've posted this issue on [Stackoverflow](https://stackoverflow.com/q/62730422/975887), but this seems like an issue with the library.

I'm trying to use Magick.NET library to do image processing on Azure Functions.

I have tried the same code in a Console App (.NET Core 3.1) which works without any issues. However when running the same code in an Azure Functions Project, I get the following error.

**Code:**

`var image = new MagickImage(File.ReadAllBytes(@"<My Image Path>"));`

**Exception:**

`System.TypeInitializationException: 'The type initializer for 'NativeMagickSettings' threw an exception.'`

**StackTrace:**

    at ImageMagick.MagickSettings.NativeMagickSettings..ctor()
    at ImageMagick.MagickSettings..ctor()
    at ImageMagick.MagickImage..ctor()
    at ImageMagick.MagickImage..ctor(Byte[] data)
    at FunctionApp2.Function1.Run(TimerInfo myTimer, ILogger log) <-- My Code

**Inner Exception:**

`Unable to load DLL 'Magick.Native-Q8-x64.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)`

**Inner Exception StackTrace**

    at ImageMagick.Environment.NativeMethods.X64.Environment_Initialize()
    at ImageMagick.Environment.NativeEnvironment.Initialize()
    at ImageMagick.Environment.Initialize()
    at ImageMagick.MagickSettings.NativeMagickSettings..cctor()

I believe this is due to the way Azure Functions package is organized and the path Magick.NET is looking for its native binary to load. I can confirm that the native binary is written at `\bin\Debug\netcoreapp3.1\runtimes\<platform>\native` when the Function app is built.

I have traced the apps using ProcessMonitor for each app and can see Function app looks in the wrong location.

It looks in 

>\bin\Debug\netcoreapp3.1\ **bin\\** runtimes\\[platform]\native 

instead of 

>\bin\Debug\netcoreapp3.1\runtimes\\[platform]\native

(Below traces are filtered to only show `path contains Magick.Native` and process name filtered to the Console app exe and func.exe)

[Console App Trace][1]
[![Console App Trace][2]][2]

[Function App Trace][3]
[![Function App Trace][4]][4]

I have also tried setting `MagickAnyCPU.CacheDirectory` to `Directory.GetCurrentDirectory()` but it hasn't helped.

After building, if I copy `\netcoreapp3.1\runtimes` folder to `\bin` this fixes the problem, but I current don't want to rely on that trick if there's a proper workaround.

So, is there a workaround to get ImageMagick to load native lib from the right place?

  [1]: https://pastebin.com/i94vDB41
  [2]: https://i.stack.imgur.com/rKlHu.png
  [3]: https://pastebin.com/FnzAhu21
  [4]: https://i.stack.imgur.com/IMnbs.png

**Comment by madushans** [comment]
**UPDATE**

This issue appears to happen only when using 
`<TargetFramework>netcoreapp3.1</TargetFramework>`
and `Microsoft.NET.Sdk.Functions` version `3.0.x`

It does not appear when using older versions. I have successfully used
`<TargetFramework>netcoreapp2.1</TargetFramework>`
and `Microsoft.NET.Sdk.Functions` version `1.0.x`
without this issue.

**Comment by madushans** [comment]
Potential workaround

Add following to csproj

```
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="xcopy /Y /E $(OutDir)runtimes\ $(OutDir)bin\runtimes\" />
  </Target>
```

or right click project > properties > Build Events and paste the above command to `post build event command line`

This will copy all of the `runtimes` binaries, not just the one we want, increasing the size of final build.

edit: updated the command to have /Y

**Comment by dlemstra** [comment, marked as answer]
This looks more like an issue with the project system itself? This library is following the standards of where the files should be in the NuGet package. You could use `MagickNET.SetNativeLibraryDirectory` as a work around for your issue instead of copying the files.

**Reply by madushans** [reply]
@dlemstra You're right. This seems to only happen when using `Microsoft.NET.Sdk.Functions` version `3.0.8` which seems to have only just appeared in nuget on Friday. I will instead downgrade to `3.0.7`.

**Reply by madushans** [reply]
**Update:** This issue was reported to [Azure Functions VS SDK ](https://github.com/Azure/azure-functions-vs-build-sdk/issues/437)and was resolved in [version 3.0.9](https://www.nuget.org/packages/Microsoft.NET.Sdk.Functions/3.0.9)

**Reply by dlemstra** [reply]
Thanks for coming back and adding the link to that issue 👍
